docs: add stream processing examples (#3916)
confluentinc/ksql: docs-md/concepts/stream-processing.md

Each time a new song is inserted into the `rock_songs` topic, the uppercase
version of the title is appended to the `title_cased_songs` stream.

Deriving a new stream from an existing table
--------------------------------------------------

Given the following table:

```sql
CREATE TABLE products (product_name VARCHAR, cost DOUBLE)
    WITH (kafka_topic='products', partitions=2, value_format='json');
```

> You can create a stream that will capture all writes made to the table:

Reviewer: ```suggestion
You can create a stream that captures all writes made to the table:
```

```sql
CREATE STREAM products_changelog AS
    SELECT * FROM products EMIT CHANGES;
```

> The `products_changelog` stream will emit a new row any time a row in the `products` table is

Reviewer: ```suggestion
The `products_changelog` stream emits a new row any time a row in the `products` table is
```

added or modified.

Deriving a new table from an existing stream
-----------------------------------------------

Given the following table and stream:

```sql
CREATE TABLE products (product_name VARCHAR, cost DOUBLE)
    WITH (kafka_topic='products', partitions=1, value_format='json', key='product_name');

CREATE STREAM orders (product_name VARCHAR)
    WITH (kafka_topic='orders', partitions=1, value_format='json', key='product_name');
```

You can create an table that aggregates rows from the `orders` stream, while also joining the stream on
the `products` table to enrich the `orders` data:

```sql
CREATE TABLE order_metrics AS
    SELECT p.product_name, COUNT(*) AS count, SUM(p.cost) AS revenue
    FROM orders o JOIN products p ON p.product_name = o.product_name
    GROUP BY p.product_name EMIT CHANGES;
```

> This aggregate table will keep track of the total number of orders of each product, along with the total amount of revenue

Reviewer: ```suggestion
This aggregate table keeps track of the total number of orders of each product, along with the total amount of revenue
```

generated by each product.

Deriving a new table from an existing table
-------------------------------------------------

Given the following aggregate table:

```sql
CREATE TABLE page_view_metrics AS
    SELECT url, location_id, COUNT(*) AS count
    FROM page_views GROUP BY url EMIT CHANGES;
```

You can derive another table that filters out rows from the `page_view_metrics` table:

```sql
CREATE TABLE page_view_metrics_mountain_view AS
    SELECT url, count FROM page_view_metrics
    WHERE location_id = 42 EMIT CHANGES;
```

Deriving a new stream from multiple streams
-------------------------------------------------

Given the following two streams:

```sql
CREATE STREAM impressions (user VARCHAR, impression_id BIGINT, url VARCHAR)
    WITH (kafka_topic='impressions', partitions=1, value_format='json', key='user');

CREATE STREAM clicks (user VARCHAR)
    WITH (kafka_topic='clicks', partitions=1, value_format='json', key='user');
```

You can create a derived stream that joins the `impressions` and `clicks` streams to output

> rows indicating that a given impression has been clicked on within one minute of the initial ad impression:

Reviewer: ```suggestion
rows indicating that a given impression has been clicked within one minute of the initial ad impression:
```

```sql
CREATE STREAM clicked_impressions AS
    SELECT * FROM impressions i JOIN clicks c WITHIN 1 minute ON i.user = c.user EMIT CHANGES;
```

Any time an `impressions` row is received, followed within one minute by a `clicks` row having the same `user`, 

> a row will be emitted into the `clicked_impressions` stream.

Reviewer: ```suggestion
a row is emitted into the `clicked_impressions` stream.
```
